guardian/typerighter
# Use Guardian NER service for entity recognition

Co-authored-by: @rhystmills 

## What does this change?

Our [first attempt](https://github.com/guardian/typerighter/pull/443) at NER (named entity recognition) has resulted in a [reduction in noise](https://github.com/guardian/typerighter/pull/443#issuecomment-1748715074) but many proper nouns remain unidentified and are therefore flagged as typos by typerighter.

![peckham](https://github.com/guardian/typerighter/assets/7423751/e0aad7b0-bd9b-478d-b201-6b9629fb4082)

This PR swaps our NER implementation from the [OpenNLP](https://opennlp.apache.org/) library to the Guardian's own NER service, which is accessed via a REST API. This results in a significant further reduction in noise. However, there are a few risks associated with this approach, as outlined below.

If we decide to go ahead with this approach, we can strip out the OpenNLP implementation in a subsequent PR. 

## How to test

Either run this branch locally (in conjunction with composer) or deploy it to CODE. Paste in some text and see how many proper nouns are flagged as spelling mistakes, compared with doing the same thing on `main`. There are a few test articles below.

## Have we considered potential risks?

There are 3 main risks:

1) As before, there is a risk with any NER solution that genuinely misspelled words might be incorrectly identified as named-entities and therefore excluded from spell-checking (i.e. false negatives).

2) There is a risk that adding this integration degrades or overwhelms the Guardian NER service because of the increased traffic to the API. This would be bad because it is used by other production systems. We can mitigate this by working closely with the data science team to monitor the service.

3) There is a risk that introducing a network request increases the latency of the checker and this feels obvious to the user.

## How can we measure success?

Fewer proper nouns being flagged as spelling mistakes. No significant hit to the usability of the tool.

## Images

### [Ukraine benchmarking article](https://composer.code.dev-gutools.co.uk/content/652cf0ba8f081a136f3a40f5)

|Before|After|
|-|-|
| ![Screenshot 2023-10-23 at 10 17 14](https://github.com/guardian/typerighter/assets/7423751/12b57c6f-edc6-4203-884b-33333f2219c2) | ![Screenshot 2023-10-23 at 10 17 36](https://github.com/guardian/typerighter/assets/7423751/759427f4-3104-4064-8c00-45dd4034f1bf) |

### [Israel benchmarking article](https://composer.code.dev-gutools.co.uk/content/652cf03d8f081a136f3a40f3)

|Before|After|
|-|-|
| ![Screenshot 2023-10-23 at 10 16 58](https://github.com/guardian/typerighter/assets/7423751/c108fe7c-0b51-4756-b6d4-29befee1b792) | ![Screenshot 2023-10-23 at 10 17 06](https://github.com/guardian/typerighter/assets/7423751/72d9af75-9414-420f-8e7b-074d1a013138) |

### [Art benchmarking article](https://composer.code.dev-gutools.co.uk/content/65363bf68f08b60f838f1eed)

|Before|After|
|-|-|
| ![Screenshot 2023-10-23 at 10 53 06](https://github.com/guardian/typerighter/assets/7423751/101054a2-4998-4754-b6a4-75e36c789c0c) | ![Screenshot 2023-10-23 at 10 53 28](https://github.com/guardian/typerighter/assets/7423751/068609a9-7247-4184-b4f1-59645e6143d5)

## Guardian NER Links

Repo: https://github.com/guardian/data-science-ner-service
Logs: https://logs.gutools.co.uk/s/data-science/app/discover 
PROD Service: https://ner.gutools.co.uk/v1/
CODE Service: https://ner.code.dev-gutools.co.uk/v1/
